service/history/statemachines/tasks.go

```diff
@@ -0,0 +1,93 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package statemachines
+
+import (
+	"errors"
+	"time"
+
+	"go.temporal.io/server/service/history/tasks"
+)
+
+// ErrInvalidTaskKind can be returned by a [TaskSerializer] if it received the wrong task kind.
+var ErrInvalidTaskKind = errors.New("invalid task kind")
+
+// Task type.
+type TaskType struct {
+	// Type ID that is used to minimize the persistence storage space and look up the regisered serializer.
+	// Type IDs are expected to be immutable as a serializer must be compatible with the task's persistent data.
+	ID int32
+	// Human readable name for this type.
+	Name string
+}
+
+// A Task is generated by a state machine in order to drive execution. For example, a callback state machine in the
+// SCHEDULED state, would generate an invocation task to be eventually executed by the framework. State machine
+// transitions and tasks are committed atomically to ensure that the system is in a consistent state.
+//
+// Tasks are generated by calling the [StateMachine.Tasks] method on a state machine after it has transitioned. Tasks
+// are executed by an executor that is registered to handle a specific task type. The framework converts this minimal
+// task representation into [tasks.Task] instances, filling in the state machine reference, workflow key, and task ID.
+// A [TaskSerializer] need to be registered in a [Registry] for a given type in order to process tasks of that type.
+type Task interface {
+	// Task type that must be unique per task definition.
+	Type() TaskType
+	// Kind of the task, see [TaskKind] for more info.
+	// Note that each category supports a specific task kind, make sure those match.
+	Kind() TaskKind
+	// Determines which queue this task is scheduled on.
+	Category() tasks.Category
```

Comment:
It's not used here, it's used by the task generator when the `tasks.Task` instances are generated.

I had a version of this where `Kind` and `Category` were merged but that seemed less flexible since we may want multiple timer or immediate task queues.

Lets leave this for now and consider merging the concepts if we see they're redundant after a few more use cases.